File: app/src/main/java/com/sg/moviesindex/db/typeconverters/GenreTypeConverter.java
package com.sg.moviesindex.db.typeconverters;

import androidx.room.TypeConverter;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.sg.moviesindex.model.tmdb.Genre;

import java.lang.reflect.Type;
import java.util.ArrayList;

public class GenreTypeConverter {

    private static final Gson gson = new Gson();

    @TypeConverter
    public static ArrayList<Genre> gettingListFromString(String data) {
        if (data == null) {
            return new ArrayList<Genre>();
        }

        Type listType = new TypeToken<ArrayList<Genre>>() {
        }.getType();

        return gson.fromJson(data, listType);
    }

    @TypeConverter
    public static String gettingStringFromList(ArrayList<Genre> someObjects) {
        return gson.toJson(someObjects);
    }

}

File: app/src/main/java/com/sg/moviesindex/data/local/typeconverters/GenreTypeConverter.kt
package com.sg.moviesindex.data.local.typeconverters

import androidx.room.TypeConverter
import com.sg.moviesindex.data.remote.Genre
import com.squareup.moshi.Moshi
import com.squareup.moshi.Types

class GenreTypeConverter {
  private val moshi = Moshi.Builder().build()
  private val listType = Types.newParameterizedType(List::class.java, Genre::class.java)
  private val adapter = moshi.adapter<List<Genre>>(listType)

  @TypeConverter
  fun gettingListFromString(data: String?): List<Genre> {
    if (data == null) {
      return emptyList()
    }
    return adapter.fromJson(data) ?: emptyList()
  }

  @TypeConverter
  fun gettingStringFromList(someObjects: List<Genre>): String = adapter.toJson(someObjects)
}
